packages/core/source-map/src/SourceMap.js

```diff
@@ -0,0 +1,385 @@
+// @flow
+import {SourceMapConsumer, SourceMapGenerator} from 'source-map';
+import lineCounter from '../../utils/src/lineCounter';
+
+type PositionType = {
+  line: number,
+  column: number
+};
+
+type MappingType = {
+  generated: PositionType,
+  original: PositionType | null,
+  source: string | null,
+  name?: string | null
+};
+
+type RawSourceMapType = {
+  version: number,
+  sources: Array<string>,
+  names: Array<string>,
+  sourceRoot?: string,
+  sourcesContent?: Array<string>,
+  mappings: string,
+  file: string
+};
+
+type RawMapInputType = SourceMapConsumer | string | RawSourceMapType;
+
+type ConsumerMappingItemType = {
+  source: string,
+  generatedLine: number,
+  generatedColumn: number,
+  originalLine: number,
+  originalColumn: number,
+  name: string
+};
+
+export default class SourceMap {
+  mappings: Array<MappingType>;
+  sources: Map<string, string | null>;
+
+  constructor(
+    mappings?: Array<MappingType> = [],
+    sources?: Map<string, string> | {[key: string]: string}
+  ) {
+    // We probably only wanna run this in dev, as sourcemaps should be fast in prod...
+    this._validateMappings(mappings);
+
+    this.mappings = mappings;
+
+    if (sources) {
+      let iteratable =
+        typeof sources === 'object' ? Object.entries(sources) : sources;
+      // $FlowFixMe
+      this.sources = new Map(iteratable);
+    } else {
+      this.sources = new Map();
+    }
+  }
+
+  _validateMappings(mappings: Array<MappingType>) {
+    for (let mapping of mappings) {
+      if (!mapping) {
+        throw new Error('mapping is undefined');
+      }
+
+      if (!mapping.generated) {
+        throw new Error('generated mapping is undefined');
+      }
+
+      if (!mapping.source) {
+        throw new Error('source should be defined');
+      }
+
+      let isValidOriginal =
+        mapping.original === null ||
+        (typeof mapping.original.line === 'number' &&
```

Comment:
Since the typing guarantees that if `original` is present, doesn't that mean that `mapping.original.line` must be a number?

Reply:
oh, I see this is all validation stuff and probably should err on the side of being safe.